diff --git a/include/network/string_view.hpp b/include/network/string_view.hpp
index 8ff7a5e..a482a51 100644
--- a/include/network/string_view.hpp
+++ b/include/network/string_view.hpp
@@ -164,8 +164,8 @@ class basic_string_view {
                ? traits::compare(data(), s.data(), size())
                : (size() < s.size()
                       ? (traits::compare(data(), s.data(), size()) > 0 ? 1 : -1)
-                      : (traits::compare(data(), s.data(), size()) < 0 ? -1
-                                                                       : 1));
+                      : (traits::compare(data(), s.data(), s.size()) < 0 ? -1
+                                                                         : 1));
   }
 
   constexpr int compare(size_type pos1, size_type n1,
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index d7f387c..231e116 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -14,13 +14,8 @@ set(Uri_SRCS
   ${CMAKE_CURRENT_SOURCE_DIR}/detail/uri_resolve.cpp
   )
 
-add_library(network-uri ${Uri_SRCS})
+add_library(network-uri STATIC ${Uri_SRCS})
 target_link_libraries(network-uri)
-if(${CMAKE_CXX_COMPILER_ID} MATCHES Clang)
-  if (NOT Uri_DISABLE_LIBCXX)
-    target_link_libraries(network-uri "c++")
-  endif()
-endif()
 
 #propagate sources to parent scope for one-lib-build
 set(Uri_SRCS ${Uri_SRCS} PARENT_SCOPE)
diff --git a/src/detail/uri_parse.cpp b/src/detail/uri_parse.cpp
index 7f77148..73f581e 100644
--- a/src/detail/uri_parse.cpp
+++ b/src/detail/uri_parse.cpp
@@ -31,7 +31,7 @@ bool validate_scheme(string_view::const_iterator &it,
   }
 
   // The first character must be a letter
-  if (!std::isalpha(*it, std::locale("C"))) {
+  if (!std::isalpha(*it, std::locale::classic())) {
     return false;
   }
   ++it;
